agent/tests/test_php_execute.c

```diff
@@ -101,20 +101,238 @@ static void test_php_cur_stack_depth(TSRMLS_D) {
+      "RuntimeException('Division by zero'); } else return $a; }");
+  tlib_php_request_eval("function two($a) { return three($a); }");
+  tlib_php_request_eval("function uncaught($a) { return two($a); }");
+  tlib_php_request_eval(
+      "function caught($a) { try {two($a);} catch (Exception $e) { return 1;} "
+      "return 1; }");
+  tlib_php_request_eval(
+      "function followup($a) { try {two($a);} catch (Exception $e) { return "
+      "three(1);} return three(1); }");
+  tlib_php_request_eval(
+      "function followup_uncaught($a) { try {two($a);} catch (Exception $e) { "
+      "return three(0);} return three(1); }");
+  tlib_php_request_eval(
+      "function rethrow($a) { try {two($a);} catch (Exception $e) { throw new "
+      "RuntimeException('Rethrown caught exception: '. $e->getMessage());} "
+      "return three(1); }");
+}
+
+static void test_stack_depth_after_exception() {
+  zval* expr = NULL;
+  zval* arg = NULL;
+
+  /*
+   * call a function and trigger an exception and cause two segments to dangle
+   * because it was caught, even though two functions don't get the oapi end
+   * func call, they are still cleaned up and stack_depth is appropriately
+   * decremented. the valid function end will trigger a cleanup dangling
+   * segments, stack_depth should be zero again
+   */
+
+  /*
+   * stack depth should increment on function call and decrement on function
+   * end.
+   */
+  tlib_php_request_start();
+  populate_functions();
+  /*
+   * pass argument that will not throw exception.
+   * stack depth should be 0 before calling and 1 before ending.
+   */
+  tlib_pass_if_int_equal(
+      "PHP stack depth tracking should be 0 before function call", 0,
+      NRPRG(php_cur_stack_depth));
+  arg = tlib_php_request_eval_expr("1" TSRMLS_CC);
+  expr = nr_php_call(NULL, "uncaught", arg);
+  tlib_pass_if_not_null("Runs fine with no exception.", expr);
+  tlib_pass_if_zval_type_is("Should have received the arg value.", IS_LONG,
+                            expr);
+  tlib_pass_if_int_equal(
+      "PHP stack depth tracking should be 0 after successful function call", 0,
+      NRPRG(php_cur_stack_depth));
+
+  nr_php_zval_free(&expr);
+  nr_php_zval_free(&arg);
+
+  /*
+   * call a function and trigger an exception and cause three segments to dangle
+   * stack_depth should be initially stuck at 3
+   * after triggering the unwind, stack_depth should be zero again
+   */
+  tlib_pass_if_int_equal(
+      "PHP stack depth tracking should be 0 before function call", 0,
+      NRPRG(php_cur_stack_depth));
+  arg = tlib_php_request_eval_expr("0");
+  expr = nr_php_call(NULL, "uncaught", arg);
+  tlib_pass_if_int_equal(
+      "PHP stack depth tracking should be 3 after function call", 3,
+      NRPRG(php_cur_stack_depth));
+  tlib_pass_if_null("Exception so expr should be null.", expr);
+
+  /*
+   * Trigger the unwind.
+   */
+  tlib_php_request_eval("newrelic_end_transaction(); ");
```

Comment:
What happens in a PHP application that doesn't call `newrelic_end_transaction();`? I guess `nr_php_txn_end` will get called eventually (at RSHUTDOWN?) which will call `nr_php_stacked_segment_unwind`, yes?

Reply:
Yes, it gets called automatically here: https://github.com/newrelic/newrelic-php-agent/blob/main/agent/php_txn.c#L1053 if a txn is detected as active at the end of rshutdown: https://github.com/newrelic/newrelic-php-agent/blob/main/agent/php_rshutdown.c#L121
Using newrelic_end_transaction was a way for me to force it to occur while still retaining request state that I could query and check for correct values.